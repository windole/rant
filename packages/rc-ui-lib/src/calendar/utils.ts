/* eslint-disable no-nested-ternary */
export function compareMonth(date1: Date, date2: Date): number {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

export function compareDay(day1: Date, day2: Date): number {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

export const cloneDate = (date: Date): Date => new Date(date);

export const cloneDates = (dates: Date | Date[]): Date | Date[] =>
  Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

export function getDayByOffset(date: Date, offset: number): Date {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

export const getPrevDay = (date: Date): Date => getDayByOffset(date, -1);
export const getNextDay = (date: Date): Date => getDayByOffset(date, 1);
export const getToday = (): Date => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

export function calcDateNum(date: [Date, Date]): number {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

const calendarLang = {
  end: '结束',
  start: '开始',
  title: '日期选择',
  confirm: '确定',
  startEnd: '开始/结束',
  weekdays: ['日', '一', '二', '三', '四', '五', '六'],
  monthTitle: (year: number, month: number) => `${year}年${month}月`,
  rangePrompt: (maxRange: number) => `最多选择 ${maxRange} 天`,
};

export const t = (key: string, ...args): string => {
  if (args.length) {
    return calendarLang[key](...args);
  }
  return calendarLang[key];
};

export const formatMonthTitle = (date: Date): string =>
  t('monthTitle', date.getFullYear(), date.getMonth() + 1);
