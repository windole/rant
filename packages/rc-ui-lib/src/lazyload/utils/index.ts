import { ListenEvent } from '../PropsType';

/* eslint-disable prefer-rest-params */
const inBrowser = typeof window !== 'undefined';

export const DEFAULT_URL =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

// 需要监听的事件
export const DEFAULT_EVENTS: ListenEvent[] = [
  'scroll',
  'wheel',
  'resize',
  'mousewheel',
  'animationend',
  'transitionend',
  'touchmove',
];

export const hasIntersectionObserver =
  inBrowser &&
  'IntersectionObserver' in window &&
  'IntersectionObserverEntry' in window &&
  'intersectionRatio' in window.IntersectionObserverEntry.prototype;

export const modeType = {
  event: 'event',
  observer: 'observer',
};

export const call = (fn): void => fn && fn();

export function setImage(
  element: HTMLMediaElement | HTMLIFrameElement | HTMLImageElement | HTMLElement,
  imagePath: string,
): void {
  if (
    element instanceof HTMLIFrameElement ||
    element instanceof HTMLImageElement ||
    element instanceof HTMLMediaElement
  ) {
    element.src = imagePath;
  } else {
    element.style.backgroundImage = `url('${imagePath}')`;
  }
}

export function loadImage(imagePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = imagePath;
    img.onload = () => resolve(imagePath);
    img.onerror = (err) => reject(err);
  });
}

export function addCssClassName(
  element: HTMLImageElement | HTMLElement,
  cssClassName: string,
): void {
  if (!element.className.includes(cssClassName)) {
    element.className += ` ${cssClassName}`;
  }
}

export function throttle(fn, threshold = 250, scope?) {
  let last;
  let deferTimer;
  return () => {
    // @ts-ignore
    const context = scope || this;

    const now = +new Date();
    const args = arguments;
    if (last && now < last + threshold) {
      // hold on to it
      clearTimeout(deferTimer);
      deferTimer = setTimeout(() => {
        last = now;
        fn.apply(context, args);
      }, threshold);
    } else {
      last = now;
      fn.apply(context, args);
    }
  };
}

export function debounce(func, wait, immediate?) {
  let timeout;
  let args;
  let context;
  let timestamp;
  let result;

  const later = function later() {
    const last = +new Date() - timestamp;

    if (last < wait && last >= 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        if (!timeout) {
          context = null;
          args = null;
        }
      }
    }
  };

  return function debounced() {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    context = this;
    args = arguments;
    timestamp = +new Date();

    const callNow = immediate && !timeout;
    if (!timeout) {
      timeout = setTimeout(later, wait);
    }

    if (callNow) {
      result = func.apply(context, args);
      context = null;
      args = null;
    }

    return result;
  };
}
