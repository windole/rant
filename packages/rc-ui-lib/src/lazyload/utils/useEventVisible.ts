import { useEffect, useRef, useState } from 'react';
import { debounce as debounceFunc, DEFAULT_EVENTS, throttle as throttleFunc } from '.';
import { EventOptions } from '../PropsType';

// try to handle passive events
let passiveEventSupported = false;
try {
  const opts = Object.defineProperty({}, 'passive', {
    get() {
      passiveEventSupported = true;
    },
  });
  window.addEventListener('test', null, opts);
  // eslint-disable-next-line no-empty
} catch (e) {}

const passiveEvent = passiveEventSupported ? { capture: false, passive: true } : false;

const useEventVisible = (
  eventOptions: EventOptions,
): [React.MutableRefObject<HTMLDivElement | HTMLImageElement>, boolean] => {
  const {
    listenEvents = DEFAULT_EVENTS,
    offset,
    scrollContainer,
    debounce,
    throttle,
  } = eventOptions;
  // 是否可见
  const [visible, setVisible] = useState(false);
  // // 包裹容器
  const containerRef = useRef<HTMLDivElement | HTMLImageElement>(null);

  useEffect(() => {
    // 检查函数
    let checkVisible = () => {
      // 如果容器不存在则不计算
      if (!containerRef.current) return;
      // 获取当前组件位置
      const { top, left, width, height } = containerRef.current?.getBoundingClientRect();
      // 获取滚动容器位置
      const {
        top: parentTop,
        left: parentLeft,
        width: parentWidth,
        height: parentHeight,
      } = scrollContainer.getBoundingClientRect();
      // 计算屏幕高度和宽度
      const windowInnerHeight = window.innerHeight || document.documentElement.clientHeight;
      const windowInnerWidth = window.innerWidth || document.documentElement.clientWidth;
      // 计算滚动容器和屏幕可视区的相交区域
      const intersectionTop = Math.min(parentTop, 0);
      const intersectionLeft = Math.min(parentLeft, 0);

      const intersectionHeight =
        Math.min(windowInnerHeight, parentTop + parentHeight) - intersectionTop;
      const intersectionWidth =
        Math.min(windowInnerWidth, parentLeft + parentWidth) - intersectionLeft;
      // 计算组件距离可视区的高度
      const offsetTop = top - intersectionTop;
      const offsetLeft = left - intersectionLeft;
      // 偏移量计算[top, left]
      const offsets = Array.isArray(offset) ? offset : [offset, offset];
      // 通过上述距离判断组件是否在可见区域
      const isVisible =
        offsetTop - offsets[0] <= intersectionHeight &&
        offsetTop + height + offsets[0] >= 0 &&
        offsetLeft - offsets[1] <= intersectionWidth &&
        offsetLeft + width + offsets[1] >= 0;

      setVisible(isVisible);
      // 如果可见取消当前组件的所有的监听
      if (isVisible) {
        listenEvents.forEach((event) => {
          scrollContainer?.removeEventListener(event, checkVisible);
        });
      }
    };

    // 使用防抖节流增强checkVisible函数，这里优先防抖，其次节流
    if (debounce) {
      // 防抖函数
      checkVisible = debounceFunc(checkVisible, debounce);
    } else if (throttle) {
      // 节流函数
      checkVisible = throttleFunc(checkVisible, throttle);
    }
    // 批量监听事件
    listenEvents.forEach((event) => {
      scrollContainer?.addEventListener(event, checkVisible, passiveEvent);
    });

    // 自动执行一次
    checkVisible();

    return () => {
      listenEvents.forEach((event) => {
        scrollContainer?.removeEventListener(event, checkVisible, passiveEvent);
      });
    };
  }, [scrollContainer]);
  return [containerRef, visible];
};

export default useEventVisible;
