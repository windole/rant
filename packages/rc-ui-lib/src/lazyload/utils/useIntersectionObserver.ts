import { useCallback, useEffect, useState } from 'react';

const useIntersectionObserver = (
  observerOptions: IntersectionObserverInit,
  handleIntersection: IntersectionObserverCallback,
): ((node: Element) => void) => {
  const [element, setElement] = useState<Element>(null);

  const handleElement = useCallback((node: Element) => {
    setElement(node);
  }, []);

  useEffect(() => {
    if (!element) return;
    const observer = new IntersectionObserver(handleIntersection, observerOptions);
    observer.observe(element);
    // eslint-disable-next-line consistent-return
    return () => observer && observer.disconnect && observer.disconnect();
  }, [handleIntersection, observerOptions, element]);
  return handleElement;
};

export default useIntersectionObserver;
