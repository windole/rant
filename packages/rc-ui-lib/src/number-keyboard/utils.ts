import { KeyConfig, NumberKeyboardProps } from './PropsType';

const genBasicKeys = (props: NumberKeyboardProps) => {
  const keys: KeyConfig[] = Array(9)
    .fill('')
    .map((_, i) => ({ text: i + 1 }));

  if (props.randomKeyOrder) {
    keys.sort(() => (Math.random() > 0.5 ? 1 : -1));
  }

  return keys;
};

export const genDefaultKeys = (props: NumberKeyboardProps): KeyConfig[] => [
  ...genBasicKeys(props),
  { text: props.extraKey as string, type: 'extra' },
  { text: 0 },
  {
    text: props.showDeleteKey ? props.deleteButtonText : '',
    type: props.showDeleteKey ? 'delete' : '',
  },
];

export const genCustomKeys = (props: NumberKeyboardProps) => {
  const keys = genBasicKeys(props);
  const { extraKey } = props;
  const extraKeys = Array.isArray(extraKey) ? extraKey : [extraKey];

  if (extraKeys.length === 1) {
    keys.push({ text: 0, wider: true }, { text: extraKeys[0], type: 'extra' });
  } else if (extraKeys.length === 2) {
    keys.push(
      { text: extraKeys[0], type: 'extra' },
      { text: 0 },
      { text: extraKeys[1], type: 'extra' },
    );
  }

  return keys;
};
