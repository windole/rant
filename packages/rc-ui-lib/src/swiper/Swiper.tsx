/* eslint-disable no-return-assign */
import React, {
  forwardRef,
  useContext,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from 'react';
import classnames from 'classnames';
import { useSpring, animated } from '@react-spring/web';
import { useDrag } from '@use-gesture/react';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import { PageIndicator } from './page-indicator';
import SwiperItem from './SwiperItem';
import { SwiperProps, SwiperInstance } from './PropsType';
import useRefState from '../hooks/use-ref-state';
import { bound } from '../utils/bound';
import { devWarning } from '../utils/dev-log';
import { noop } from '../utils';
import { getRect } from '../hooks/get-rect';
import useMountedRef from '../hooks/use-mounted-ref';
import { mergeFuncProps } from '../utils/with-func-props';

function modulus(value: number, division: number) {
  const remainder = value % division;
  return remainder < 0 ? remainder + division : remainder;
}

const eventToPropRecord = {
  mousedown: 'onMouseDown',
  mousemove: 'onMouseMove',
  mouseup: 'onMouseUp',
} as const;

let currentUid: undefined | {};

const Swiper = forwardRef<SwiperInstance, SwiperProps>((props, ref) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('swiper', prefixCls);

  const { loop: outerLoop, autoplay, direction, autoplayInterval } = props;

  const [uid] = useState({});

  const lock = useRef<boolean>(false);
  const trackRef = useRef<HTMLDivElement>(null);
  const [root, setRoot] = useState<HTMLDivElement>(null);
  const [current, setCurrent] = useState(props.defaultIndex);
  const [dragging, setDragging, draggingRef] = useRefState(false);
  const isVertical = direction === 'vertical';

  const axis = useMemo(() => (isVertical ? 'y' : 'x'), [isVertical]);

  const slideRatio = props.slideSize / 100;
  const offsetRatio = props.trackOffset / 100;

  const computedStyle = useMemo(() => {
    return {
      [`--${prefixCls}-swipe-slide-size`]: `${props.slideSize}%`,
      [`--${prefixCls}-swipe-track-offset`]: `${props.trackOffset}%`,
      ...props.style,
    };
  }, [props.style, props.slideSize, props.trackOffset]);

  const axisDistance = useMemo(() => {
    if (!isVertical) return 100;
    const rect = getRect(root);
    return rect.height;
  }, [isVertical, root]);

  const { validChildren, count } = useMemo(() => {
    let innerCount = 0;
    const innerValidChildren = React.Children.map(props.children, (child) => {
      if (!React.isValidElement(child)) return null;
      if (child.type !== SwiperItem) {
        devWarning('Swiper', 'The children of `Swiper` must be `Swiper.Item` components.');
        return null;
      }
      innerCount++;
      return child;
    });
    return {
      validChildren: innerValidChildren,
      count: innerCount,
    };
  }, [props.children]);

  const loop = useMemo(() => {
    if (count <= 1) return false;
    return outerLoop;
  }, [count, outerLoop]);

  const mountedRef = useMountedRef();

  const getSlidePixels = () => {
    const track = trackRef.current;
    if (!track) return 0;
    const trackPixels = isVertical ? track.offsetHeight : track.offsetWidth;
    return (trackPixels * props.slideSize) / 100;
  };

  const boundIndex = (cur: number) => {
    let min = 0;
    let max = count - 1;
    if (props.stuckAtBoundary) {
      min += (1 - slideRatio - offsetRatio) / slideRatio;
      max -= (1 - slideRatio - offsetRatio) / slideRatio;
    }
    return bound(cur, min, max);
  };

  const [{ position }, api] = useSpring(
    () => ({
      position: boundIndex(current) * 100,
      config: {
        tension: 200,
        friction: 30,
        autoplayInterval,
      },
      onRest: () => {
        if (draggingRef.current) return;
        const rawX = position.get();
        const totalWidth = 100 * count;
        const standardPosition = modulus(rawX, totalWidth);
        if (standardPosition === rawX) return;
        api.start({
          position: standardPosition,
          immediate: true,
        });
      },
    }),
    [count],
  );

  const dragCancelRef = useRef<(() => void) | null>(null);
  function forceCancelDrag() {
    dragCancelRef.current?.();
    draggingRef.current = false;
  }

  const bind = useDrag(
    (state) => {
      dragCancelRef.current = state.cancel;
      if (!state.intentional) return;
      if (state.first && !currentUid) {
        currentUid = uid;
      }
      if (currentUid !== uid) return;
      currentUid = state.last ? undefined : uid;
      const slidePixels = getSlidePixels();

      if (lock.current) return;
      if (!slidePixels) return;
      const paramIndex = isVertical ? 1 : 0;
      const offset = state.offset[paramIndex];
      const itemDirection = state.direction[paramIndex];
      const velocity = state.velocity[paramIndex];
      setDragging(true);
      if (!state.last) {
        api.start({
          position: (offset * 100) / slidePixels,
          immediate: true,
        });
      } else {
        const index = Math.round(
          (offset + Math.min(velocity * 2000, slidePixels) * itemDirection) / slidePixels,
        );
        swipeTo(index);
        window.setTimeout(() => {
          if (mountedRef.current) {
            setDragging(false);
          }
        });
      }
    },
    {
      transform: ([x, y]) => [-x, -y],
      from: () => {
        const slidePixels = getSlidePixels();
        return [(position.get() / 100) * slidePixels, (position.get() / 100) * slidePixels];
      },
      bounds: () => {
        if (loop) return {};
        const slidePixels = getSlidePixels();
        const lowerBound = boundIndex(0) * slidePixels;
        const upperBound = boundIndex(count - 1) * slidePixels;
        return isVertical
          ? {
              top: lowerBound,
              bottom: upperBound,
            }
          : {
              left: lowerBound,
              right: upperBound,
            };
      },
      rubberband: true,
      axis,
      preventScroll: !isVertical,
      pointer: {
        touch: true,
      },
    },
  );

  const renderIndicator = () => {
    if (props.indicator === undefined || props.indicator === true) {
      return (
        <div className={classnames(bem('indicator', { vertical: isVertical }))}>
          <PageIndicator
            {...props.indicatorProps}
            direction={props.direction}
            total={count}
            current={current}
          />
        </div>
      );
    }
    if (typeof props.indicator === 'function') {
      return props.indicator(count, current);
    }
    return null;
  };

  const onClickCapture = (e) => {
    if (draggingRef.current) {
      e.stopPropagation();
    }
    forceCancelDrag();
  };

  function swipeTo(index: number, immediate = false) {
    if (loop) {
      const i = modulus(index, count);
      setCurrent(i);
      props.onIndexChange?.(i);
      api.start({
        position: index * 100,
        immediate,
      });
    } else {
      const i = bound(index, 0, count - 1);
      setCurrent(i);
      props.onIndexChange?.(i);
      api.start({
        position: boundIndex(i) * 100,
        immediate,
      });
    }
  }

  const swipeNext = () => {
    swipeTo(Math.round(position.get() / 100) + 1);
  };

  const swipePrev = () => {
    swipeTo(Math.round(position.get() / 100) - 1);
  };

  useImperativeHandle(ref, () => ({
    activeIndex: current,
    swipeTo,
    swipeNext,
    swipePrev,
    lock: () => {
      lock.current = true;
    },
    unlock: () => {
      lock.current = false;
    },
  }));

  useEffect(() => {
    if (!autoplay || dragging) return noop;
    const interval = window.setInterval(() => {
      swipeNext();
    }, autoplayInterval);
    return () => {
      window.clearInterval(interval);
    };
  }, [autoplay, dragging, axisDistance]);

  if (count === 0) {
    devWarning('Swiper', '`Swiper` needs at least one child.');
  }

  const dragProps = { ...(props.allowTouchMove ? bind() : {}) };

  const stopPropagationProps: Partial<Record<any, any>> = {};
  props.stopPropagation.forEach(key => {
    const prop = eventToPropRecord[key];
    stopPropagationProps[prop] = function (e: Event) {
      e.stopPropagation();
    };
  })


  const mergedProps = mergeFuncProps(dragProps, stopPropagationProps);

  return (
    <div
      ref={setRoot}
      className={classnames(props.className, bem({ vertical: isVertical }))}
      style={computedStyle}
    >
      <div
        ref={trackRef}
        className={classnames(
          bem('track', {
            'allow-touch-move': props.allowTouchMove,
          }),
        )}
        onClickCapture={onClickCapture}
        {...mergedProps}
      >
        <div
          className={classnames(
            bem('track-inner', {
              vertical: isVertical,
            }),
          )}
        >
          {React.Children.map(validChildren, (child, index) => {
            return (
              <animated.div
                className={classnames(bem('slide'))}
                style={{
                  [axis]: position.to((pos) => {
                    let finalPosition = -pos + index * 100;
                    const totalWidth = count * 100;
                    const flagWidth = totalWidth / 2;
                    finalPosition = modulus(finalPosition + flagWidth, totalWidth) - flagWidth;
                    return `${finalPosition}%`;
                  }),
                  left: `-${index * 100}%`,
                }}
              >
                {child}
              </animated.div>
            );
          })}
        </div>
      </div>
      {renderIndicator()}
    </div>
  );
});

Swiper.defaultProps = {
  autoplay: false,
  autoplayInterval: 3000,
  allowTouchMove: true,
  defaultIndex: 0,
  loop: true,
  slideSize: 100,
  stuckAtBoundary: false,
  trackOffset: 0,
  stopPropagation: [],
};

Swiper.displayName = 'Swiper';

export default Swiper;
