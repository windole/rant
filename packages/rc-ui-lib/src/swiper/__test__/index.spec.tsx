/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, cleanup, waitFor, act } from '@testing-library/react';
import TestsEvent from '../../../tests/events';
import { sleep } from '../../../tests/utils';
import Swiper, { SwiperInstance, SwiperProps } from '..';

interface ElementRect {
  bottom: number;
  height: number;
  left: number;
  right: number;
  top: number;
  width: number;
  x: number;
  y: number;
}

const swipeStyle = {
  width: '100px',
  height: '100px',
};

const $props = {
  loop: true,
  touchable: true,
  autoplay: false,
  initialSwipe: 0,
  allowTouchMove: true,
};

describe('Swipe test with testing library', () => {
  let eachClears: Array<() => void> = [];

  const mockGetBoundingClientRect = (rect: Partial<ElementRect>): (() => void) => {
    const originMethod = Element.prototype.getBoundingClientRect;

    Element.prototype.getBoundingClientRect = jest.fn(
      () =>
        ({
          top: 0,
          left: 0,
          right: 0,
          bottom: 0,
          width: 0,
          height: 0,
          ...rect,
        } as any),
    );

    return () => {
      Element.prototype.getBoundingClientRect = originMethod;
    };
  };

  const mockOffset = (c: Element) => {
    Object.defineProperty(c, 'offsetHeight', {
      configurable: true,
      get: () => 100,
    });
    Object.defineProperty(c, 'offsetWidth', {
      configurable: true,
      get: () => 100,
    });
  };

  function createSwipe(props: SwiperProps) {
    const swipeRef = React.createRef<SwiperInstance>();

    const { queryByTestId, container, rerender, debug } = render(
      <Swiper ref={swipeRef} style={swipeStyle} {...props}>
        <Swiper.Item>1</Swiper.Item>
        <Swiper.Item>2</Swiper.Item>
        <Swiper.Item>3</Swiper.Item>
      </Swiper>,
    );

    return {
      container,
      rerender,
      swipeRef,
      debug,
      queryByTestId,
    };
  }

  beforeEach(() => {
    jest.useFakeTimers();

    const clear = mockGetBoundingClientRect({
      width: 100,
      height: 100,
    });

    eachClears.push(clear);
  });

  afterEach(() => {
    eachClears.forEach((clear) => clear());
    eachClears = [];

    cleanup();
    jest.restoreAllMocks();
  });

  it('should swipe to specific swipe after calling the swipeTo method', async () => {
    const onIndexChange = jest.fn();

    const { swipeRef } = createSwipe({
      ...$props,
      onIndexChange,
    });

    const swipeTo = (step: number) => {
      swipeRef.current.swipeTo(step);
    };

    await waitFor(async () => {
      swipeTo(2);
    });

    await sleep(100);

    expect(onIndexChange).toHaveBeenCalledWith(2);
  });

  it('should swipe to next swipe after calling next method', async () => {
    const onIndexChange = jest.fn();

    const { swipeRef } = createSwipe({
      ...$props,
      onIndexChange,
    });

    await waitFor(async () => {
      swipeRef.current.swipeNext();
    });

    await sleep(100);

    expect(onIndexChange).toHaveBeenCalledWith(1);
  });

  it('should not swipe to next swipe after calling unlock method', async () => {
    const onIndexChange = jest.fn();

    const { container, swipeRef } = createSwipe({
      ...$props,
      onIndexChange,
    });

    const track = container.querySelector('.rc-swiper__track');
    mockOffset(track);

    act(() => {
      swipeRef.current.lock();
    });

    await TestsEvent.triggerDrag(track, [-100, 0]);

    expect(onIndexChange).not.toHaveBeenCalled();

    act(() => {
      swipeRef.current.unlock();
    });

    await TestsEvent.triggerDrag(track, [-100, 0]);

    expect(onIndexChange).toHaveBeenCalled();
  });

  it('should swipe to prev swipe after calling prev method', async () => {
    const onIndexChange = jest.fn();
    const { swipeRef } = createSwipe({
      ...$props,
      onIndexChange,
    });

    act(() => {
      swipeRef.current.swipePrev();
    });

    await sleep(100);

    expect(onIndexChange).toHaveBeenCalledWith(2);
  });

  it('should render initial swipe correctly', async () => {
    const { container, rerender } = createSwipe({
      ...$props,
    });

    expect(container).toMatchSnapshot();

    const props2 = {
      ...$props,
      initialSwipe: 2,
    };

    rerender(
      <Swiper style={swipeStyle} {...props2}>
        <Swiper.Item>1</Swiper.Item>
        <Swiper.Item>2</Swiper.Item>
        <Swiper.Item>3</Swiper.Item>
      </Swiper>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly when prop stuckAtBoundary is true', async () => {
    const onIndexChange = jest.fn();

    const { container } = createSwipe({
      ...$props,
      loop: false,
      stuckAtBoundary: true,
      onIndexChange,
    });

    const track = container.querySelector('.rc-swiper__track');
    mockOffset(track);
    await TestsEvent.triggerDrag(track, [100, 0]);
    await sleep(100);
    expect(onIndexChange).toHaveBeenLastCalledWith(0);

    await sleep(1000);
    expect(container).toMatchSnapshot();
  });

  it('should render vertical swipe when using vertical prop', async () => {
    const onIndexChange = jest.fn();
    const { container } = createSwipe({
      ...$props,
      direction: 'vertical',
      allowTouchMove: true,
      onIndexChange,
    });

    const track = container.querySelector('.rc-swiper__track');
    mockOffset(track);

    await TestsEvent.triggerDrag(track, [0, -100]);

    await sleep(1000);
    expect(onIndexChange).not.toHaveBeenCalledWith(0);
  });

  it('should render swipe looply when using loop prop', async () => {
    const onIndexChange = jest.fn();
    const { container } = createSwipe({
      ...$props,
      loop: true,
      onIndexChange,
    });

    const track = container.querySelector('.rc-swiper__track');
    mockOffset(track);
    await TestsEvent.triggerDrag(track, [-100, 0]);
    await sleep(100);
    expect(onIndexChange).not.toHaveBeenLastCalledWith(0);
  });

  it('should autoplay when using autoplay prop', async () => {
    const onIndexChange = jest.fn();
    createSwipe({
      ...$props,
      autoplay: true,
      autoplayInterval: 80,
      onIndexChange,
    });

    act(() => {
      jest.advanceTimersByTime(200);
    });

    expect(onIndexChange).toHaveBeenCalled();
  });

  it('should render correctly when Swiper not child', async () => {
    const spyConsole: jest.SpyInstance = jest.spyOn(console, 'warn');
    spyConsole.mockImplementation(() => {
      return null;
    });
    const { container } = render(<Swiper style={swipeStyle} {...$props} />);

    expect(container).toMatchSnapshot();
    expect(spyConsole).toHaveBeenCalledWith('[Swiper] `Swiper` needs at least one child.');
    spyConsole.mockRestore();
  });

  it('should render correctly when using customer renderIndicator', async () => {
    const { container } = createSwipe({
      ...$props,
      indicator: (total, current) => (
        <div className="custom-indicator">
          {current + 1}/{total}
        </div>
      ),
    });

    expect(container).toMatchSnapshot();
  });

  it('stop propagation should be work', async () => {
    const onMouseDown = jest.fn();
    const onMouseMove = jest.fn();
    const onMouseUp = jest.fn();

    const { container } = render(
      <Swiper style={swipeStyle} stopPropagation={['mousedown', 'mousemove', 'mouseup']}>
        <Swiper.Item>1</Swiper.Item>
        <Swiper.Item>2</Swiper.Item>
        <Swiper.Item>3</Swiper.Item>
      </Swiper>,
    );

    const track = container.querySelector('.rc-swiper__track');
    mockOffset(track);
    await TestsEvent.triggerDrag(track, [-100, 0]);
    await sleep(100);

    expect(onMouseDown).not.toBeCalled();
    expect(onMouseMove).not.toBeCalled();
    expect(onMouseUp).not.toBeCalled();
  });
});
