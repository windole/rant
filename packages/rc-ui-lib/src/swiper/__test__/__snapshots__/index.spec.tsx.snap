// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swipe test with testing library should render correctly when Swiper not child 1`] = `
<div>
  <div
    class="rc-swiper"
    style="--rc-swipe-slide-size: 100%; --rc-swipe-track-offset: 0%; width: 100px; height: 100px;"
  >
    <div
      class="rc-swiper__track rc-swiper__track--allow-touch-move"
    >
      <div
        class="rc-swiper__track-inner"
      />
    </div>
    <div
      class="rc-swiper__indicator"
    >
      <div
        class="rc-indicator"
      />
    </div>
  </div>
</div>
`;

exports[`Swipe test with testing library should render correctly when prop stuckAtBoundary is true 1`] = `
<div>
  <div
    class="rc-swiper"
    style="--rc-swipe-slide-size: 100%; --rc-swipe-track-offset: 0%; width: 100px; height: 100px;"
  >
    <div
      class="rc-swiper__track rc-swiper__track--allow-touch-move"
    >
      <div
        class="rc-swiper__track-inner"
      >
        <div
          class="rc-swiper__slide"
          style="left: -0%; transform: none;"
        >
          <div
            class="rc-swiper-item"
          >
            1
          </div>
        </div>
        <div
          class="rc-swiper__slide"
          style="left: -100%; transform: translate3d(100%,0,0);"
        >
          <div
            class="rc-swiper-item"
          >
            2
          </div>
        </div>
        <div
          class="rc-swiper__slide"
          style="left: -200%; transform: translate3d(-100%,0,0);"
        >
          <div
            class="rc-swiper-item"
          >
            3
          </div>
        </div>
      </div>
    </div>
    <div
      class="rc-swiper__indicator"
    >
      <div
        class="rc-indicator"
      >
        <div
          class="rc-indicator__dot rc-indicator__dot--active"
        />
        <div
          class="rc-indicator__dot"
        />
        <div
          class="rc-indicator__dot"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Swipe test with testing library should render correctly when using customer renderIndicator 1`] = `
<div>
  <div
    class="rc-swiper"
    style="--rc-swipe-slide-size: 100%; --rc-swipe-track-offset: 0%; width: 100px; height: 100px;"
  >
    <div
      class="rc-swiper__track rc-swiper__track--allow-touch-move"
    >
      <div
        class="rc-swiper__track-inner"
      >
        <div
          class="rc-swiper__slide"
          style="left: -0%; transform: none;"
        >
          <div
            class="rc-swiper-item"
          >
            1
          </div>
        </div>
        <div
          class="rc-swiper__slide"
          style="left: -100%; transform: translate3d(100%,0,0);"
        >
          <div
            class="rc-swiper-item"
          >
            2
          </div>
        </div>
        <div
          class="rc-swiper__slide"
          style="left: -200%; transform: translate3d(-100%,0,0);"
        >
          <div
            class="rc-swiper-item"
          >
            3
          </div>
        </div>
      </div>
    </div>
    <div
      class="custom-indicator"
    >
      1
      /
      3
    </div>
  </div>
</div>
`;

exports[`Swipe test with testing library should render initial swipe correctly 1`] = `
<div>
  <div
    class="rc-swiper"
    style="--rc-swipe-slide-size: 100%; --rc-swipe-track-offset: 0%; width: 100px; height: 100px;"
  >
    <div
      class="rc-swiper__track rc-swiper__track--allow-touch-move"
    >
      <div
        class="rc-swiper__track-inner"
      >
        <div
          class="rc-swiper__slide"
          style="left: -0%; transform: none;"
        >
          <div
            class="rc-swiper-item"
          >
            1
          </div>
        </div>
        <div
          class="rc-swiper__slide"
          style="left: -100%; transform: translate3d(100%,0,0);"
        >
          <div
            class="rc-swiper-item"
          >
            2
          </div>
        </div>
        <div
          class="rc-swiper__slide"
          style="left: -200%; transform: translate3d(-100%,0,0);"
        >
          <div
            class="rc-swiper-item"
          >
            3
          </div>
        </div>
      </div>
    </div>
    <div
      class="rc-swiper__indicator"
    >
      <div
        class="rc-indicator"
      >
        <div
          class="rc-indicator__dot rc-indicator__dot--active"
        />
        <div
          class="rc-indicator__dot"
        />
        <div
          class="rc-indicator__dot"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Swipe test with testing library should render initial swipe correctly 2`] = `
<div>
  <div
    class="rc-swiper"
    style="--rc-swipe-slide-size: 100%; --rc-swipe-track-offset: 0%; width: 100px; height: 100px;"
  >
    <div
      class="rc-swiper__track rc-swiper__track--allow-touch-move"
    >
      <div
        class="rc-swiper__track-inner"
      >
        <div
          class="rc-swiper__slide"
          style="left: -0%; transform: none;"
        >
          <div
            class="rc-swiper-item"
          >
            1
          </div>
        </div>
        <div
          class="rc-swiper__slide"
          style="left: -100%; transform: translate3d(100%,0,0);"
        >
          <div
            class="rc-swiper-item"
          >
            2
          </div>
        </div>
        <div
          class="rc-swiper__slide"
          style="left: -200%; transform: translate3d(-100%,0,0);"
        >
          <div
            class="rc-swiper-item"
          >
            3
          </div>
        </div>
      </div>
    </div>
    <div
      class="rc-swiper__indicator"
    >
      <div
        class="rc-indicator"
      >
        <div
          class="rc-indicator__dot rc-indicator__dot--active"
        />
        <div
          class="rc-indicator__dot"
        />
        <div
          class="rc-indicator__dot"
        />
      </div>
    </div>
  </div>
</div>
`;
